spikeinterface/core/waveform_extractor.py

```diff
@@ -64,10 +64,10 @@ def __init__(self, recording, sorting, folder):
 
         # cache in memory
         self._waveforms = {}
-        self._template_std = {}
-        self._template_average = {}
-        self._template_median = {}
-        self._template_quantile = {}
+        self._template_std = {i:dict() for i in self.sorting.unit_ids}
+        self._template_average = {i:dict() for i in self.sorting.unit_ids}
+        self._template_median = {i:dict() for i in self.sorting.unit_ids}
+        self._template_quantile = {i:dict() for i in self.sorting.unit_ids}
```

Comment:
@Saksham20 in my opinion, `templates` should be defined over the entire recording, not by segments. Imagine this scenario: you have a 2 segment object, with 15 minutes for the first template and only 2 minutes for the second. In this case, the template will be well defined on the first recording and it might be noisy on the second due to the small number of spikes. Then why one would compute the template on the second segment only? 

The second problem is caching. The way it is done in this PR, if a user first computes the template on the second segment and then `get_template` for the first one, the second template will be returned since it's already cached. 

Of course it's always possible to compute templates by segment, by instantiating a `WaveformExtractor` object for each segment, but this is a particular use case. As for the NWB conversion, I think that the same template mean and std should be written to all unit tables.

Reply:
>The second problem is caching. The way it is done in this PR, if a user first computes the template on the second segment and then get_template for the first one, the second template will be returned since it's already cached.

Yes, this was a bug, I've fixed it here 
https://github.com/SpikeInterface/spikeinterface/pull/258/commits/429a66087ac1fed7c6a29eec378fe0cab1aa90eb

>templates should be defined over the entire recording, not by segments. Imagine this scenario: you have a 2 segment object, with 15 minutes for the first template and only 2 minutes for the second. In this case, the template will be well defined on the first recording and it might be noisy on the second due to the small number of spikes. Then why one would compute the template on the second segment only?

if a user wants to get templates for the entire recording, simply omiting the argument `segment_index` takes care of it. The key assumed for a segment_index for 'all' segments is -1. This value is also cached. 
[Check here](https://github.com/SpikeInterface/spikeinterface/pull/258/files#diff-d917025308798740765907787da9eeb8d9658b090bb95cd02847a264fecbc656R226)

Reply:
Maybe a `segment_index=None` by default then?

Reply:
I kept it that so that it can directly be a key to the cache dict.

Reply:
I see. IMO it makes sense to have to different variables for caching: e.g. `self._template_average` and `self._template_average_by_segment`.

@samuelgarcia what do you think?